src/session/session_api.c

```diff
@@ -920,6 +921,57 @@ err:	F_CLR(session, WT_SESSION_CAN_WAIT | WT_SESSION_NO_CACHE_CHECK);
 }
 
 /*
+ * __session_snapshot --
+ *	WT_SESSION->snapshot method.
+ */
+static int
+__session_snapshot(WT_SESSION *wt_session, const char *config)
+{
+	WT_DECL_RET;
+	WT_SESSION_IMPL *session;
+	WT_TXN *txn;
+
+	session = (WT_SESSION_IMPL *)wt_session;
+
+	txn = &session->txn;
+
+	SESSION_API_CALL(session, snapshot, config, cfg);
+
+	/*
+	 * Checkpoints require a snapshot to write a transactionally consistent
+	 * snapshot of the data.
+	 *
+	 * We can't use an application's transaction: if it has uncommitted
+	 * changes, they will be written in the checkpoint and may appear after
+	 * a crash.
+	 *
+	 * Use a real snapshot transaction: we don't want any chance of the
+	 * snapshot being updated during the checkpoint.  Eviction is prevented
+	 * from evicting anything newer than this because we track the oldest
+	 * transaction ID in the system that is not visible to all readers.
+	 */
+	if (F_ISSET(txn, WT_TXN_RUNNING))
+		WT_ERR_MSG(session, EINVAL,
+		    "Snapshot not permitted in a transaction");
+
+	/*
+	 * Reset open cursors.  Do this explicitly, even though it will happen
+	 * implicitly in the call to begin_transaction for the checkpoint, the
+	 * checkpoint code will acquire the schema lock before we do that, and
+	 * some implementation of WT_CURSOR::reset might need the schema lock.
+	 */
+	WT_ERR(__wt_session_reset_cursors(session));
+
+	/* Drop any snapshots to be removed first. */
+	WT_ERR(__wt_txn_named_snapshot_drop(session, cfg));
+
+	/* Start the named snapshot if requested. */
+	WT_ERR(__wt_txn_named_snapshot_begin(session, cfg));
```

Comment:
This separation will drop `nsnap_rwlock` in between the drop and the begin.  While I don't think applications should call this method simultaneously from multiple threads, if they do I think we need to keep the lock across both calls to avoid unexpected results.